### PR TITLE
feat: render selection and cursor.

### PR DESCRIPTION
**Pull request by LucasXu0**
#633

### CODE REVIEW

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart [hunk 1]
```diff
@@ -26,7 +29,37 @@ class _ImageNodeWidget extends StatelessWidget {
...
+}
+
+class __ImageNodeWidgetState extends State<_ImageNodeWidget> with Selectable {
+  Node get node => widget.node;
+  EditorState get editorState => widget.editorState;
+  String get src => widget.node.attributes['image_src'] as String;
+
+  @override
+  List<Rect> getSelectionRectsInSelection(Offset start, Offset end) {
+    final renderBox = context.findRenderObject() as RenderBox;
```
Comment (vincentdchan) [inline frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart:43]: What if the `RenderObject` is not a `RenderBox`?

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart [hunk 2]
```diff
@@ -26,7 +29,37 @@ class _ImageNodeWidget extends StatelessWidget {
...
+    final size = renderBox.size;
+    final boxOffset = renderBox.localToGlobal(Offset.zero);
+    return [boxOffset & size];
+  }
+
+  @override
+  Rect getCursorRect(Offset start) {
+    final renderBox = context.findRenderObject() as RenderBox;
+    final size = Size(5, renderBox.size.height);
```
Comment (vincentdchan) [inline frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart:52]: Maybe using a constant variable to store the magic number is better.

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/selected_text_node_widget.dart [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  @override
+  TextSelection? getTextSelection() {
+    return _textSelection;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    print('text rebuild $this');
```
Comment (vincentdchan) [inline frontend/app_flowy/packages/flowy_editor/example/lib/plugin/selected_text_node_widget.dart:102]: `debugPrint`

File: frontend/app_flowy/packages/flowy_editor/lib/document/node.dart [hunk 4]
```diff
@@ -184,8 +186,6 @@ class TextNode extends Node {
     return map;
   }
 
-  String toRawString() => _delta.operations
-      .whereType<TextInsert>()
-      .map((op) => op.content)
-      .toString();
+  String toRawString() =>
```
Comment (vincentdchan) [inline frontend/app_flowy/packages/flowy_editor/lib/document/node.dart:189]: The selection request will this method frequently. Actually the result can be cached.

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart (hunk 1 above)
Comment (LucasXu0) [reply frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart:43]: Depends on the user's computation. It must be a `RenderBox` in this situation.

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart (hunk 2 above)
Comment (LucasXu0) [reply frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart:52]: Yes, but the image_node_widget is just an example, so I use a number directly.

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/selected_text_node_widget.dart (hunk 3 above)
Comment (LucasXu0) [reply frontend/app_flowy/packages/flowy_editor/example/lib/plugin/selected_text_node_widget.dart:102]: OK

File: frontend/app_flowy/packages/flowy_editor/lib/document/node.dart (hunk 4 above)
Comment (LucasXu0) [reply frontend/app_flowy/packages/flowy_editor/lib/document/node.dart:189]: As a TODO. have added a comment.

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart (hunk 1 above)
Comment (appflowy) [reply frontend/app_flowy/packages/flowy_editor/example/lib/plugin/image_node_widget.dart:54]: Documenting it in your code would be better.

File: frontend/app_flowy/packages/flowy_editor/example/lib/plugin/selected_text_node_widget.dart (hunk 3 above)
Comment (appflowy) [reply frontend/app_flowy/packages/flowy_editor/example/lib/plugin/selected_text_node_widget.dart:102]: > `debugPrint`

How to disable the debug print if the user doesn't want to print the log. Is there something like env print?

**Pull request merged**